x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformScheduler.java

```diff
@@ -0,0 +1,388 @@
+        Event event = new Event(scheduledTask.getTransformId(), scheduledTask.getNextScheduledTimeMillis(), currentTimeMillis);
+        // Trigger the listener
+        scheduledTask.getListener().triggered(event);
+        // Update the task's last_triggered_time to current time (next_scheduled_time gets automatically re-calculated)
+        scheduledTasks.update(scheduledTask.getTransformId(), task -> {
+            if (task.equals(scheduledTask) == false) {
+                logger.debug(
+                    new ParameterizedMessage(
+                        "[{}] task object got modified while processing. Expected: {}, was: {}",
+                        scheduledTask.getTransformId(),
+                        scheduledTask,
+                        task
+                    )
+                );
+            }
+            return new ScheduledTransformTask(
+                task.getTransformId(),
+                task.getFrequency(),
+                currentTimeMillis,
+                task.getFailureCount(),
+                task.getListener()
+            );
+        });
+    }
+
+    /**
+     * Stops the processing.
+     */
+    public void stop() {
+        threadPool.shutdown();
+    }
+
+    /**
+     * Registers the transform by adding it to the queue.
+     *
+     * @param transformTaskParams structure containing transform's id and frequency
+     * @param listener listener to be triggered
+     */
+    public void registerTransform(TransformTaskParams transformTaskParams, Listener listener) {
+        String transformId = transformTaskParams.getId();
+        logger.trace(new ParameterizedMessage("[{}] register the transform", transformId));
+        long currentTimeMillis = clock.millis();
+        ScheduledTransformTask scheduledTransformTask = new ScheduledTransformTask(
+            transformId,
+            transformTaskParams.getFrequency(),
+            null,  // this task has not been triggered yet
+            0,  // this task has not failed yet
+            currentTimeMillis,  // we schedule this task at current clock time so that it is processed ASAP
+            listener
+        );
+        scheduledTasks.add(scheduledTransformTask);
+        processScheduledTasks();
+    }
+
+    /**
+     * Updates the transform task's failure count.
+     * Updating the failure count affects the task's next_scheduled_time and may result in the task being processed earlier that it would
+     * normally (i.e.: according to its frequency) be.
+     *
+     * @param transformId id of the transform to update
+     * @param failureCount new value of transform task's failure count
+     */
+    public void handleTransformFailureCountChanged(String transformId, int failureCount) {
+        logger.trace(new ParameterizedMessage("[{}] handle transform failure count change to {}", transformId, failureCount));
+        // Update the task's failure count (next_scheduled_time gets automatically re-calculated)
+        scheduledTasks.update(
```

Comment:
The `Map<String, E> elementsById;` in `ConcurrentPriorityQueue`, is the only purpose of that for updating the elements?

Can you drop the map from `ConcurrentPriorityQueue` and do a lookup in the sorted set for the `ScheduledTransformTask`?  Then just remove the old element and put the newly computed one.